cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -819,6 +832,378 @@ const FunctionDoc map_lookup_doc{
+          } else {
+            builder->UnsafeAppendNull();
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<InputType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Fixed-size binary types, including decimals
+  template <typename InputType>
+  std::enable_if_t<is_fixed_size_binary_type<InputType>::value, Status> Visit(
+      const InputType& input_type, KernelContext* ctx, const ExecSpan& batch,
+      ExecResult* out) {
+    using BuilderType = typename TypeTraits<InputType>::BuilderType;
+    using ListBuilderType = typename TypeTraits<OutputType>::BuilderType;
+    auto builder = std::make_shared<BuilderType>(input_type.GetSharedPtr(),
+                                                 ctx->exec_context()->memory_pool());
+    ListBuilderType list_builder(ctx->exec_context()->memory_pool(), builder, list_type);
+
+    RETURN_NOT_OK(builder->Reserve(batch.num_values() * batch.length));
+    RETURN_NOT_OK(ReserveBinaryData<InputType>(batch, builder.get()));
+
+    RETURN_NOT_OK(list_builder.Reserve(batch.length));
+
+    for (int i = 0; i < batch.length; ++i) {
+      RETURN_NOT_OK(list_builder.Append());
+      for (const auto& input : batch.values) {
+        if (input.is_array()) {
+          const auto& arr = input.array;
+          if (arr.IsValid(i)) {
+            std::string_view view(arr.buffers[1].data_as<char>() +
+                                      (i + arr.offset) * input_type.byte_width(),
+                                  input_type.byte_width());
+            builder->UnsafeAppend(view);
+          } else {
+            builder->UnsafeAppendNull();
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<InputType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Deal with nested/union types with a naive approach: First concatenate the inputs,
+  // then shuffle it using Take
+  Status Visit(const DataType& input_type, KernelContext* ctx, const ExecSpan& batch,
+               ExecResult* out) {
+    std::vector<std::shared_ptr<ArrayData>> inputs;
+    inputs.reserve(batch.num_values());
+    // Starting index of each input in the concatenated array
+    std::vector<int64_t> input_start_index;
+    input_start_index.reserve(batch.num_values());
+    int64_t cur_index = 0;
+    for (const auto& input : batch.values) {
+      input_start_index.push_back(cur_index);
+      if (input.is_array()) {
+        inputs.emplace_back(input.array.ToArrayData());
+        cur_index += input.array.length;
+      } else {
+        ARROW_ASSIGN_OR_RAISE(auto arr_from_scalar,
+                              MakeArrayFromScalar(*input.scalar, 1));
+        inputs.emplace_back(std::move(arr_from_scalar)->data());
+        cur_index += 1;
+      }
+    }
+    ARROW_ASSIGN_OR_RAISE(auto concatenated_inputs, Concatenate(inputs));
```

Comment:
Instead of concatenating, perhaps we can just make a chunked array of the inputs?

Reply:
Right, updated